File: src/main/java/com/protryon/jasm/stage1/CConstant.java
package com.protryon.jasm.stage1;

import com.protryon.jasm.*;
import com.protryon.jasm.Method;
import com.protryon.jasm.stage1.constant.*;
import com.shapesecurity.functional.F;
import com.shapesecurity.functional.data.Either;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public abstract class CConstant<T> {

    public final T value;
    public final int type;

    protected CConstant(T value, int type) {
        this.value = value;
        this.type = type;
    }

    protected abstract void write(DataOutputStream out) throws IOException;

    protected static CConstant read(DataInputStream in) throws IOException {
        int type = in.read();
        switch (type) {
            case 1:
                byte[] utf = new byte[in.readUnsignedShort()];
                in.readFully(utf);
                return new CUTF8(new String(utf));
            case 3:
                return new CInteger(in.readInt());
            case 4:
                return new CFloat(in.readFloat());
            case 5:
                return new CLong(in.readLong());
            case 6:
                return new CDouble(in.readDouble());
            case 7:
                return new CClass(in.readUnsignedShort());
            case 8:
                return new CString(in.readUnsignedShort());
            case 9:
                return new CFieldRef(in.readUnsignedShort(), in.readUnsignedShort());
            case 10:
                return new CMethodRef(in.readUnsignedShort(), in.readUnsignedShort());
            case 11:
                return new CInterfaceMethodRef(in.readUnsignedShort(), in.readUnsignedShort());
            case 12:
                return new CNameAndType(in.readUnsignedShort(), in.readUnsignedShort());
            case 15:
                return new CMethodHandle((byte) in.read(), in.readUnsignedShort());
            case 16:
                return new CMethodType(in.readUnsignedShort());
            case 17:
                return new CDynamic(in.readUnsignedShort(), in.readUnsignedShort());
            case 18:
                return new CInvokeDynamic(in.readUnsignedShort(), in.readUnsignedShort());
            case 19:
                return new CModule(in.readUnsignedShort());
            case 20:
                return new CPackage(in.readUnsignedShort());
            default:
                throw new RuntimeException("Invalid constant type: " + type);
        }
    }

    private Method findMethod(Klass klass, String name, MethodDescriptor descriptor) {
        for (Method method : klass.methods) {
            if (method.name.equals(name) && method.descriptor.equals(descriptor)) {
                return method;
            }
        }
        if (klass.extending != null) {
            Method method = findMethod(klass.extending, name, descriptor);
            if (method != null) {
                return method;
            }
        }
        for (Klass implementing : klass.interfaces) {
            Method method = findMethod(implementing, name, descriptor);
            if (method != null) {
                return method;
            }
        }
        Method method = new Method(name, descriptor);
        method.isDummy = true;
        klass.methods.add(method);
        return method;
    }

    private Field findField(Klass klass, String name, JType type) {
        Field field = klass.fields.get(name);
        if (field != null) {
            return field;
        }
        if (klass.extending != null) {
            field = findField(klass.extending, name, type);
            if (field != null) {
                return field;
            }
        }
        for (Klass implementing : klass.interfaces) {
            field = findField(implementing, name, type);
            if (field != null) {
                return field;
            }
        }
        field = new Field(type, name);
        field.isDummy = true;
        klass.fields.put(name, field);
        return field;
    }

    protected Constant toConstant(Classpath classpath, F<Integer, Constant> lookup) {
        NameAndType nameAndType;
        switch (this.type) {
            case 1:
                return new Constant<>(((CUTF8) this).value);
            case 3:
                return new Constant<>(((CInteger) this).value);
            case 4:
                return new Constant<>(((CFloat) this).value);
            case 5:
                return new Constant<>(((CLong) this).value);
            case 6:
                return new Constant<>(((CDouble) this).value);
            case 7:
                return new Constant<>(classpath.loadKlass((String) lookup.apply(((CClass) this).value).value));
            case 8:
                return new Constant<>((String) lookup.apply(((CString) this).value).value);
            case 9:
                CFieldRef fieldRef = (CFieldRef) this;
                Klass fieldKlass = (Klass) lookup.apply(fieldRef.value.left).value;
                nameAndType = (NameAndType) lookup.apply(fieldRef.value.right).value;
                Field field = findField(fieldKlass, nameAndType.name, nameAndType.type.left().fromJust());
                if (field == null) {
                    throw new RuntimeException("Field not found: " + nameAndType.toString());
                }
                /*if (!field.type.equals(nameAndType.type.left().fromJust())) {
                    throw new RuntimeException("Type mismatch in field reference");
                }*/
                return new Constant<>(field);
            case 10: {
                CMethodRef methodRef = (CMethodRef) this;
                Klass methodKlass = (Klass) lookup.apply(methodRef.value.left).value;
                nameAndType = (NameAndType) lookup.apply(methodRef.value.right).value;
                Method method = findMethod(methodKlass, nameAndType.name, nameAndType.type.right().fromJust());
                if (method == null) {
                    throw new RuntimeException("Method descriptor not found: " + nameAndType.toString());
                }
                return new Constant<>(method);
            }
            case 11: {
                CInterfaceMethodRef methodRef = (CInterfaceMethodRef) this;
                Klass methodKlass = (Klass) lookup.apply(methodRef.value.left).value;
                nameAndType = (NameAndType) lookup.apply(methodRef.value.right).value;
                Method method = findMethod(methodKlass, nameAndType.name, nameAndType.type.right().fromJust());
                if (method == null) {
                    throw new RuntimeException("Interface method descriptor not found: " + nameAndType.toString());
                }
                return new Constant<>(method);
            }
            case 12:
                CNameAndType cNameAndType = (CNameAndType) this;
                String name = (String) lookup.apply(cNameAndType.value.left).value;
                String rawDescriptor = (String) lookup.apply(cNameAndType.value.right).value;
                if (rawDescriptor.startsWith("(")) {
                    return new Constant<>(new NameAndType(name, Either.right(MethodDescriptor.fromString(classpath, rawDescriptor))));
                }
                return new Constant<>(new NameAndType(name, Either.left(JType.fromDescriptor(classpath, rawDescriptor))));
            case 15:
                CMethodHandle handle = (CMethodHandle) this;
                return new Constant<>(new MethodHandle(lookup.apply(handle.value.right), MethodHandle.MethodHandleType.values()[handle.value.left]));
            case 16:
                CMethodType methodType = (CMethodType) this;
                return new Constant<>(MethodDescriptor.fromString(classpath, (String) lookup.apply(methodType.value).value));
            case 17:
                throw new UnsupportedOperationException("CDynamic");
                //return new CDynamic(in.readUnsignedShort(), in.readUnsignedShort());
            case 18:
                return new Constant<>((CInvokeDynamic) this); // TODO: implement
                // throw new UnsupportedOperationException("CInvokeDynamic");
            case 19:
                throw new UnsupportedOperationException("CModule");
                //return new CModule(in.readUnsignedShort());
            case 20:
                throw new UnsupportedOperationException("CPackage");
                //return new CPackage(in.readUnsignedShort());
            default:
                throw new UnsupportedOperationException("Invalid constant type: " + type);
        }
    }

    public boolean isDoubled() {
        return false;
    }
}

File: src/main/java/com/protryon/jasm/stage1/CConstant.kt
package com.protryon.jasm.stage1

import com.protryon.jasm.*
import com.protryon.jasm.Method
import com.protryon.jasm.stage1.constant.*
import com.shapesecurity.functional.data.Either

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

abstract class CConstant<T> protected constructor(val value: T, val type: Int) {

    open val isDoubled: Boolean
        get() = false

    @Throws(IOException::class)
    abstract fun write(out: DataOutputStream)

    private fun findMethod(klass: Klass, name: String, descriptor: MethodDescriptor): Method? {
        for (method in klass.methods) {
            if (method.name == name && method.descriptor == descriptor) {
                return method
            }
        }
        if (klass.extending != null) {
            val method = findMethod(klass.extending!!, name, descriptor)
            if (method != null) {
                return method
            }
        }
        for (implementing in klass.interfaces) {
            val method = findMethod(implementing, name, descriptor)
            if (method != null) {
                return method
            }
        }
        val method = Method(klass, name, descriptor)
        method.isDummy = true
        klass.methods.add(method)
        return method
    }

    private fun findField(klass: Klass, name: String, type: JType): Field? {
        var field: Field? = klass.fields[name]
        if (field != null) {
            return field
        }
        if (klass.extending != null) {
            field = findField(klass.extending!!, name, type)
            if (field != null) {
                return field
            }
        }
        for (implementing in klass.interfaces) {
            field = findField(implementing, name, type)
            if (field != null) {
                return field
            }
        }
        field = Field(klass, type, name)
        field.isDummy = true
        klass.fields[name] = field
        return field
    }

    fun toConstant(classpath: Classpath, klass: Stage1Class, lookup: (Int)->Constant<*>): Constant<*> {
        val nameAndType: NameAndType
        when (this.type) {
            1 -> return Constant((this as CUTF8).value)
            3 -> return Constant((this as CInteger).value)
            4 -> return Constant((this as CFloat).value)
            5 -> return Constant((this as CLong).value)
            6 -> return Constant((this as CDouble).value)
            7 -> {
                val descriptor = lookup.invoke((this as CClass).value).value as String
                return if (descriptor.startsWith("[")) {
                    Constant(JType.fromDescriptor(classpath, descriptor))
                } else {
                    Constant(JType.instance(classpath.loadKlass(descriptor)))
                }
            }
            8 -> return Constant(lookup.invoke((this as CString).value).value as String)
            9 -> {
                val fieldRef = this as CFieldRef
                val fieldKlassAsType = lookup.invoke(fieldRef.value.left).value as JType
                nameAndType = lookup.invoke(fieldRef.value.right).value as NameAndType
                val fieldKlass = if (fieldKlassAsType is JType.JTypeInstance) fieldKlassAsType.referenceOf() else classpath.loadKlass("java/lang/Object")
                val field = findField(fieldKlass!!, nameAndType.name, nameAndType.type.left().fromJust()) ?: throw RuntimeException("Field not found: $nameAndType")
/*if (!field.type.equals(nameAndType.type.left().fromJust())) {
                    throw new RuntimeException("Type mismatch in field reference");
                }*/
                return Constant(field)
            }
            10 -> {
                val methodRef = this as CMethodRef
                val methodKlassAsType = lookup.invoke(methodRef.value.left).value as JType
                nameAndType = lookup.invoke(methodRef.value.right).value as NameAndType
                val methodKlass = if (methodKlassAsType is JType.JTypeInstance) methodKlassAsType.referenceOf() else classpath.loadKlass("java/lang/Object")
                val method = findMethod(methodKlass!!, nameAndType.name, nameAndType.type.right().fromJust()) ?: throw RuntimeException("Method descriptor not found: $nameAndType")
                return Constant(method)
            }
            11 -> {
                val methodRef = this as CInterfaceMethodRef
                val methodKlassAsType = lookup.invoke(methodRef.value.left).value as JType
                nameAndType = lookup.invoke(methodRef.value.right).value as NameAndType
                val methodKlass = if (methodKlassAsType is JType.JTypeInstance) methodKlassAsType.referenceOf() else classpath.loadKlass("java/lang/Object")
                val method = findMethod(methodKlass!!, nameAndType.name, nameAndType.type.right().fromJust()) ?: throw RuntimeException("Interface method descriptor not found: $nameAndType")
                return Constant(method)
            }
            12 -> {
                val cNameAndType = this as CNameAndType
                val name = lookup.invoke(cNameAndType.value.left).value as String
                val rawDescriptor = lookup.invoke(cNameAndType.value.right).value as String
                return if (rawDescriptor.startsWith("(")) {
                    Constant(NameAndType(name, Either.right(MethodDescriptor.fromString(classpath, rawDescriptor)!!)))
                } else Constant(NameAndType(name, Either.left(JType.fromDescriptor(classpath, rawDescriptor))))
            }
            15 -> {
                val handle = this as CMethodHandle
                return Constant(MethodHandle(lookup.invoke(handle.value.right), MethodHandle.MethodHandleType.values()[handle.value.left.toInt()]))
            }
            16 -> {
                val methodType = this as CMethodType
                return Constant(MethodDescriptor.fromString(classpath, lookup.invoke(methodType.value).value as String))
            }
            17, 18 -> {
                val invokeDynamic = this as CInvokeDynamic
                val bootstrapMethod = klass.bootstrapMethods[invokeDynamic.value.left]!!
                val methodHandle = lookup.invoke(bootstrapMethod.methodRef)
                val arguments = Array(bootstrapMethod.arguments.size) {
                    lookup.invoke(bootstrapMethod.arguments[it])
                }
                val newBootstrapMethod = BootstrapMethod(methodHandle.value as MethodHandle, arguments)
                val bootstrapNameAndType = lookup.invoke(invokeDynamic.value.right)
                return Constant(Dynamic(this.type == 18, newBootstrapMethod, bootstrapNameAndType.value as NameAndType))
            }
            19 -> throw UnsupportedOperationException("CModule")
            //return new CModule(in.readUnsignedShort());
            20 -> throw UnsupportedOperationException("CPackage")
            //return new CPackage(in.readUnsignedShort());
            else -> throw UnsupportedOperationException("Invalid constant type: $type")
        }
    }

    companion object {

        @Throws(IOException::class)
        fun read(inputStream: DataInputStream): CConstant<*> {
            val type = inputStream.read()
            when (type) {
                1 -> {
                    val utf = ByteArray(inputStream.readUnsignedShort())
                    inputStream.readFully(utf)
                    return CUTF8(String(utf))
                }
                3 -> return CInteger(inputStream.readInt())
                4 -> return CFloat(inputStream.readFloat())
                5 -> return CLong(inputStream.readLong())
                6 -> return CDouble(inputStream.readDouble())
                7 -> return CClass(inputStream.readUnsignedShort())
                8 -> return CString(inputStream.readUnsignedShort())
                9 -> return CFieldRef(inputStream.readUnsignedShort(), inputStream.readUnsignedShort())
                10 -> return CMethodRef(inputStream.readUnsignedShort(), inputStream.readUnsignedShort())
                11 -> return CInterfaceMethodRef(inputStream.readUnsignedShort(), inputStream.readUnsignedShort())
                12 -> return CNameAndType(inputStream.readUnsignedShort(), inputStream.readUnsignedShort())
                15 -> return CMethodHandle(inputStream.read().toByte(), inputStream.readUnsignedShort())
                16 -> return CMethodType(inputStream.readUnsignedShort())
                17 -> return CDynamic(inputStream.readUnsignedShort(), inputStream.readUnsignedShort())
                18 -> return CInvokeDynamic(inputStream.readUnsignedShort(), inputStream.readUnsignedShort())
                19 -> return CModule(inputStream.readUnsignedShort())
                20 -> return CPackage(inputStream.readUnsignedShort())
                else -> throw RuntimeException("Invalid constant type: $type")
            }
        }
    }
}
